syntax = "proto3";

package aea.eightballer.hackathon_process.v1_0_0;

message HackathonProcessMessage{

  // Custom Types
  message ErrorCode{
    enum ErrorCodeEnum {
      ERROR_CODE_ENUM_NOT_FOUND = 0;
      ERROR_CODE_ENUM_INVALID_SUBMISSION = 1;
      ERROR_CODE_ENUM_INTERNAL_ERROR = 2;
      ERROR_CODE_ENUM_PERMISSION_DENIED = 3;
    }
    ErrorCodeEnum error_code = 1;
  }

  message SubmissionStatus{
    enum SubmissionStatusEnum {
      SUBMISSION_STATUS_ENUM_PENDING = 0;
      SUBMISSION_STATUS_ENUM_APPROVED = 1;
      SUBMISSION_STATUS_ENUM_REJECTED = 2;
    }
    SubmissionStatusEnum status = 1;
  }

  message SubmissionType{
    enum SubmissionTypeEnum {
      SUBMISSION_TYPE_ENUM_IDEA = 0;
      SUBMISSION_TYPE_ENUM_UPDATE = 1;
      SUBMISSION_TYPE_ENUM_FINAL = 2;
    }
    SubmissionTypeEnum submission_type = 1;
  }


  // Performatives and contents
  message Idea_Proposal_Performative{
    string team_id = 1;
    string idea_title = 2;
    string idea_description = 3;
  }

  message Idea_Feedback_Performative{
    string team_id = 1;
    string feedback = 2;
    bool approved = 3;
  }

  message Project_Update_Performative{
    string team_id = 1;
    string update_description = 2;
  }

  message Final_Submission_Performative{
    string team_id = 1;
    string project_repo = 2;
    string presentation_link = 3;
    bool presentation_link_is_set = 4;
  }

  message Submission_Acknowledgement_Performative{
    string team_id = 1;
    SubmissionType submission_type = 2;
    SubmissionStatus status = 3;
  }

  message Error_Performative{
    ErrorCode error_code = 1;
    string error_msg = 2;
  }


  oneof performative{
    Error_Performative error = 5;
    Final_Submission_Performative final_submission = 6;
    Idea_Feedback_Performative idea_feedback = 7;
    Idea_Proposal_Performative idea_proposal = 8;
    Project_Update_Performative project_update = 9;
    Submission_Acknowledgement_Performative submission_acknowledgement = 10;
  }
}
